lib/ansible/executor/process/connection.py

```diff
@@ -0,0 +1,204 @@
+                    signal.alarm(self.connection._play_context.timeout)
+
+                    start = datetime.datetime.now()
+                    resp = self.srv.handle_request(data)
+                    end = datetime.datetime.now()
+
+                    signal.alarm(0)
+
+                    display.vvvv('request took %s seconds to execute' % (end - start), host=self.connection._play_context.remote_addr)
+                    send_data(s, to_bytes(resp))
+
+                s.close()
+
+        except Exception as e:
+            # socket.accept() will raise EINTR if the socket.close() is called
+            if hasattr(e, 'errno'):
+                if e.errno != errno.EINTR:
+                    display.debug(traceback.format_exc())
+            else:
+                display.vvvv(str(e), host=self.connection._play_context.remote_addr)
+                display.debug(traceback.format_exc())
+                raise AnsibleError(str(e))
+
+        finally:
+            # when done, close the connection properly and cleanup
+            # the socket file so it can be recreated
+            self.shutdown()
+
+    def connect_timeout(self, signum, frame):
+        display.debug('persistent connection idle timeout triggered, timeout value is %s secs' % C.PERSISTENT_CONNECT_TIMEOUT)
+        self.shutdown()
+
+    def request_timeout(self, signum, frame):
+        display.debug('command timeout triggered, timeout value is %s secs' % self.connection._play_context.timeout)
+        self.shutdown()
+
+    def shutdown(self):
+        """ Shuts down the local domain socket
+        """
+        display.debug('starting shutdown of persistent connection socket %s [%s]' % (self.connection.socket_path, self.connection._play_context.remote_addr))
+
+        if not os.path.exists(self.connection.socket_path):
+            display.debug('persistent connection is not active, nothing to do!')
+            return
+
+        try:
+            if self.socket:
+                display.debug('closing local listener')
+                self.socket.close()
+
+            if self.connection:
+                display.debug('closing the connection')
+                self.connection.close()
+        except:
+            pass
+
+        finally:
+            if os.path.exists(self.connection.socket_path):
+                display.debug('removing the local control socket')
+                os.remove(self.connection.socket_path)
+                setattr(self.connection, '_socket_path', None)
+                setattr(self.connection, '_connected', False)
+
+        display.debug('shutdown complete')
+
+    def start(self):
+        """ Called to initiate the connect to the remote device
+        """
+        play_context = self.connection._play_context
+
+        ssh = connection_loader.get('ssh', class_only=True)
+        cp = ssh._create_control_path(play_context.remote_addr, play_context.port, play_context.remote_user)
```

Comment:
As some network os support multiple connections type eg. junos support both network_cli and netconf, I think connection name should also be passed to _create_control_path()